File: previous_versions/octave/m-files/multiplier_calc.m
function M = multiplier_calc(line, data_spacing)

% modified from multiplier_calc_new_v1.m to use the simpler formula of Mh
% when slope > 0.45 and in the local separation zone. But the results are
% same as multiplier_calc_new.m since when Mh was calculated > 1.71 it will
% reset to 1.71 anyway.
%
% Lin 2004

% Modified slightly by Chris Thomas 2009, so that it operates only on the line of data
% passed to it rather than a whole array. Did not alter functionality

  %n = length(line);
  %[nr nc] = size(line);
  %printf('line has size %d %d\n',nr,nc);

%initialise M to 1
  M = ones(size(line));

  fwd_line = floor(line);
  % Get the index of the ridges & valleys
  ridge_ind  = findpeaks(fwd_line); % relative ind
  valley_ind = findvalleys(fwd_line); % relative ind

  if isempty(ridge_ind) % the DEM is completely flat
    %printf('Flat line\n');
    %fflush(stdout);
  elseif length(ridge_ind) == 1 & ridge_ind(1) == 1 % the DEM is downward slope all the time
    %printf('Downward slope\n');
    %fflush(stdout);
  else % 2 general cases
    if ridge_ind(1) == 1 % (1) down up down up ....
      for i=2:length(ridge_ind)
        M = max(M, Mh(fwd_line,ridge_ind(i),valley_ind(i-1),data_spacing));
      endfor
    else % (2) up dowm up dowm ....
      for i=1:length(ridge_ind)
        M = max(M, Mh(fwd_line,ridge_ind(i),valley_ind(i),data_spacing));
      endfor
    endif
  endif

endfunction

File: previous_versions/topo_python/py-files/multiplier_calc.py
##------------------------------------------------------
## Computes the multipliers for a data line
##------------------------------------------------------
import numpy
import Mh            # calculate Mh using the simpler formula modified by C.Thomas 2009        
import findpeaks     # get the indices of the ridges in a data line
import findvalleys   # get the indices of the valleys in a data line 

def multiplier_calc(line, data_spacing):
# --------------------------------------------------------
# initialise M as an array filled with 1
# --------------------------------------------------------
  nrow = numpy.size(line)
  M = numpy.ones((nrow,1),dtype=float)

  # take the largest integer of each element of the data line
  fwd_line = numpy.floor(line)

  # Get the indices of the ridges & valleys
  ridge_ind  = findpeaks.findpeaks(fwd_line)        # relative ind
  valley_ind = findvalleys.findvalleys(fwd_line)    # relative ind
 
  if numpy.size(ridge_ind) == 0: # the DEM is completely flat
     print "Flat line"
     
  elif numpy.size(ridge_ind) == 1 and ridge_ind[0] == 0:  # the DEM is downward slope all the time
     print "Downward slope"
    
  else:                        # 2 general cases, calculate m, works as Mh.m
             
      if ridge_ind[0] == 0:    # (1) down up down up ....
         for i in range(1,numpy.size(ridge_ind)):
            m = Mh.Mh(fwd_line,ridge_ind[i],valley_ind[i-1],data_spacing)
            M = numpy.maximum(M, m)
      
      else:                    # (2) up dowm up dowm ....
         for i in range(0,numpy.size(ridge_ind)):
            m = Mh.Mh(fwd_line,ridge_ind[i],valley_ind[i],data_spacing)
            M = numpy.maximum(M, m)

  return M
